compressors/typical_set_coder.py

```diff
@@ -0,0 +1,208 @@
+"""Typical set coder
+
+For a given iid source $X_1,X_2,...$ i.i.d. $\sim X$ (from alphabet $\mathcal{X}$), 
+and for given values of $n > 0$ and $\epsilon > 0$, the typical set is defined as 
+$$A_{\epsilon}^{(n)} = \left\{x^n | \left|-\frac{1}{n}\log P(x^n) - H(X)\right| \leq \epsilon\right\}$$
+
+This code works by dividing the input into blocks of size n, and then encodes each block
+as follows (note: a single data_block contains several of these n-length blocks):
+- if $x^n \in A_{\epsilon}^{(n)}$, encode as a $0$ followed by a 
+$\lceil\log_2|A_{\epsilon}^{(n)}|\rceil$ bit representation of the index of $x^n$ in the typical set
+- otherwise, encode as $1$ followed by a $\lceil\log_2|\mathcal{X}|\rceil$ representation 
+of the index of $x^n$ in $\mathcal{X}^n$
+"""
+
+from core.data_block import DataBlock
+from core.data_encoder_decoder import DataDecoder, DataEncoder
+from utils.bitarray_utils import BitArray, uint_to_bitarray, bitarray_to_uint
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from core.prob_dist import ProbabilityDist
+import numpy as np
+import itertools
+import random
+
+
+def compute_normalized_negative_log_prob_block(block, prob_dist: ProbabilityDist):
+    """Compute the normalized log probability for a block"""
+    log_prob = 0
+    for symbol in block:
+        log_prob += -np.log2(prob_dist.probability(symbol))
+    return log_prob / len(block)
+
+
+def is_typical(block, prob_dist: ProbabilityDist, eps):
+    """Check if a block is typical given a distribution and value of epsilon"""
+    if np.abs(compute_normalized_negative_log_prob_block(block, prob_dist) - prob_dist.entropy) <= eps:
```

Comment:
just `return (np.abs(compute_normalized_negative_log_prob_block(block, prob_dist) - prob_dist.entropy) <= eps)` might be good

Reply:
done